File: api/src/main/java/com/ruby/rubia_server/core/service/ZApiActivationService.java
package com.ruby.rubia_server.core.service;

import com.ruby.rubia_server.core.entity.ZApiStatus;
import com.ruby.rubia_server.core.entity.QrCodeResult;
import com.ruby.rubia_server.core.entity.PhoneCodeResult;
import com.ruby.rubia_server.core.entity.WhatsAppInstance;
import com.ruby.rubia_server.core.util.CompanyContextUtil;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.Map;

@Service
@Slf4j
public class ZApiActivationService {
    
    @Value("${zapi.clientToken}")
    private String clientToken;

    private final RestTemplate restTemplate;
    private final WhatsAppInstanceService whatsAppInstanceService;
    private final CompanyContextUtil companyContextUtil;

    public ZApiActivationService(WhatsAppInstanceService whatsAppInstanceService, 
                                CompanyContextUtil companyContextUtil) {
        this.restTemplate = new RestTemplate();
        this.whatsAppInstanceService = whatsAppInstanceService;
        this.companyContextUtil = companyContextUtil;
    }

    /**
     * Gets the active WhatsApp instance for the current company context.
     * During activation, looks for instances in configuration state.
     */
    private WhatsAppInstance getActiveInstance() {
        try {
            // First try to find a connected instance
            var connectedInstance = whatsAppInstanceService.findActiveConnectedInstance(companyContextUtil.getCurrentCompany());
            if (connectedInstance.isPresent()) {
                return connectedInstance.get();
            }
            
            // If no connected instance, look for instances in configuration/activation state
            var instances = whatsAppInstanceService.findByCompany(companyContextUtil.getCurrentCompany());
            return instances.stream()
                .filter(instance -> instance.getInstanceId() != null && instance.getAccessToken() != null)
                .filter(instance -> instance.getStatus().name().matches("CONFIGURING|AWAITING_QR_SCAN|CONNECTING"))
                .findFirst()
                .orElseThrow(() -> new IllegalStateException("No configured WhatsApp instance found for activation"));
        } catch (Exception e) {
            log.error("Error getting active instance: {}", e.getMessage());
            throw new IllegalStateException("Cannot determine active WhatsApp instance", e);
        }
    }

    /**
     * Builds the Z-API URL for the given endpoint using active instance data
     */
    private String buildZApiUrl(String endpoint) {
        WhatsAppInstance instance = getActiveInstance();
        if (instance.getInstanceId() == null || instance.getAccessToken() == null) {
            throw new IllegalStateException("WhatsApp instance is not properly configured (missing instanceId or accessToken)");
        }
        return String.format("https://api.z-api.io/instances/%s/token/%s/%s", 
                           instance.getInstanceId(), instance.getAccessToken(), endpoint);
    }

    public ZApiStatus getInstanceStatus() {
        try {
            String url = buildZApiUrl("status");
            
            HttpHeaders headers = createHeaders();
            HttpEntity<String> request = new HttpEntity<>(headers);

            ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.GET, request, Map.class);

            if (response.getStatusCode().is2xxSuccessful() && response.getBody() != null) {
                return parseStatusResponse(response.getBody());
            } else {
                return ZApiStatus.error("Failed to get status");
            }

        } catch (Exception e) {
            log.error("Error getting Z-API status: {}", e.getMessage(), e);
            return ZApiStatus.error("Error: " + e.getMessage());
        }
    }

    public QrCodeResult getQrCodeBytes() {
        try {
            String url = buildZApiUrl("qr-code");
            
            HttpHeaders headers = createHeaders();
            HttpEntity<String> request = new HttpEntity<>(headers);

            ResponseEntity<byte[]> response = restTemplate.exchange(url, HttpMethod.GET, request, byte[].class);

            if (response.getStatusCode().is2xxSuccessful() && response.getBody() != null) {
                return QrCodeResult.success(response.getBody(), "bytes");
            } else {
                return QrCodeResult.error("Failed to get QR code bytes");
            }

        } catch (Exception e) {
            log.error("Error getting QR code bytes: {}", e.getMessage(), e);
            return QrCodeResult.error("Error: " + e.getMessage());
        }
    }

    public QrCodeResult getQrCodeImage() {
        try {
            String url = buildZApiUrl("qr-code/image");
            log.info("Getting QR code from URL: {}", url);
            
            HttpHeaders headers = createHeaders();
            HttpEntity<String> request = new HttpEntity<>(headers);

            ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.GET, request, Map.class);
            log.info("Z-API QR Code response: status={}, body={}", response.getStatusCode(), response.getBody());

            if (response.getStatusCode().is2xxSuccessful() && response.getBody() != null) {
                // Z-API returns the QR code in the "value" field, not "image"
                String fullValue = (String) response.getBody().get("value");
                log.info("Full value received: {}", fullValue != null ? fullValue.substring(0, Math.min(100, fullValue.length())) + "..." : "null");
                
                // Extract base64 part from data URL format (data:image/png;base64,XXXX)
                String base64Image = null;
                if (fullValue != null && fullValue.startsWith("data:image/png;base64,")) {
                    base64Image = fullValue.substring("data:image/png;base64,".length());
                }
                
                log.info("Base64 image extracted: {}", base64Image != null ? "present (" + base64Image.length() + " chars)" : "null");
                return QrCodeResult.success(base64Image, "base64");
            } else {
                log.warn("Failed to get QR code: status={}, body={}", response.getStatusCode(), response.getBody());
                return QrCodeResult.error("Failed to get QR code image");
            }

        } catch (Exception e) {
            log.error("Error getting QR code image: {}", e.getMessage(), e);
            return QrCodeResult.error("Error: " + e.getMessage());
        }
    }

    public PhoneCodeResult getPhoneCode(String phoneNumber) {
        try {
            String url = buildZApiUrl("phone-code/" + phoneNumber);
            
            HttpHeaders headers = createHeaders();
            HttpEntity<String> request = new HttpEntity<>(headers);

            ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.GET, request, Map.class);

            if (response.getStatusCode().is2xxSuccessful() && response.getBody() != null) {
                String code = (String) response.getBody().get("code");
                return PhoneCodeResult.success(code, phoneNumber);
            } else {
                return PhoneCodeResult.error("Failed to get phone code");
            }

        } catch (Exception e) {
            log.error("Error getting phone code: {}", e.getMessage(), e);
            return PhoneCodeResult.error("Error: " + e.getMessage());
        }
    }

    public boolean restartInstance() {
        try {
            String url = buildZApiUrl("restart");
            
            HttpHeaders headers = createHeaders();
            HttpEntity<String> request = new HttpEntity<>(headers);

            ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.GET, request, Map.class);

            return response.getStatusCode().is2xxSuccessful();

        } catch (Exception e) {
            log.error("Error restarting instance: {}", e.getMessage(), e);
            return false;
        }
    }

    public boolean disconnectInstance() {
        try {
            String url = buildZApiUrl("disconnect");
            
            HttpHeaders headers = createHeaders();
            HttpEntity<String> request = new HttpEntity<>(headers);

            ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.GET, request, Map.class);

            return response.getStatusCode().is2xxSuccessful();

        } catch (Exception e) {
            log.error("Error disconnecting instance: {}", e.getMessage(), e);
            return false;
        }
    }

    private HttpHeaders createHeaders() {
        HttpHeaders headers = new HttpHeaders();
        headers.set("client-token", clientToken);
        headers.setContentType(MediaType.APPLICATION_JSON);
        return headers;
    }

    private ZApiStatus parseStatusResponse(Map<String, Object> response) {
        Object connectedObj = response.get("connected");
        Object smartphoneConnectedObj = response.get("smartphoneConnected");
        Object sessionObj = response.get("session");
        
        // Handle both boolean and string responses
        boolean connected = parseBoolean(connectedObj);
        boolean smartphoneConnected = parseBoolean(smartphoneConnectedObj);
        String session = parseString(sessionObj);
        
        return ZApiStatus.builder()
            .connected(connected)
            .session(session)
            .smartphoneConnected(smartphoneConnected)
            .needsQrCode(!connected)
            .rawResponse(response)
            .build();
    }
    
    private boolean parseBoolean(Object value) {
        if (value instanceof Boolean) {
            return (Boolean) value;
        } else if (value instanceof String) {
            return "true".equalsIgnoreCase((String) value);
        }
        return false;
    }
    
    private String parseString(Object value) {
        if (value == null) {
            return null;
        }
        return value.toString();
    }
}

File: api-dotnet/Rubia.Server/Services/ZApiActivationService.cs
using Microsoft.Extensions.Configuration;
using Rubia.Server.Entities;
using System.Text.Json;

namespace Rubia.Server.Services;

public class ZApiActivationService
{
    private readonly HttpClient _httpClient;
    private readonly WhatsAppInstanceService _whatsAppInstanceService;
    private readonly CompanyContextService _companyContextService;
    private readonly IConfiguration _configuration;
    private readonly ILogger<ZApiActivationService> _logger;

    public ZApiActivationService(
        HttpClient httpClient,
        WhatsAppInstanceService whatsAppInstanceService,
        CompanyContextService companyContextService,
        IConfiguration configuration,
        ILogger<ZApiActivationService> logger)
    {
        _httpClient = httpClient;
        _whatsAppInstanceService = whatsAppInstanceService;
        _companyContextService = companyContextService;
        _configuration = configuration;
        _logger = logger;
    }

    private async Task<WhatsAppInstance> GetActiveInstanceAsync()
    {
        try
        {
            var currentCompany = _companyContextService.GetCurrentCompany();
            
            // First try to find a connected instance
            var connectedInstance = await _whatsAppInstanceService.FindActiveConnectedInstanceAsync(currentCompany.Id);
            if (connectedInstance != null)
            {
                return connectedInstance;
            }

            // If no connected instance, look for instances that are configured
            var instances = await _whatsAppInstanceService.FindByCompanyAsync(currentCompany.Id);
            var configuredInstance = instances.FirstOrDefault(instance => 
                !string.IsNullOrEmpty(instance.InstanceId) && 
                !string.IsNullOrEmpty(instance.AccessToken) &&
                instance.IsConfigured());

            if (configuredInstance == null)
            {
                throw new InvalidOperationException("No configured WhatsApp instance found for activation");
            }

            return configuredInstance;
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error getting active instance: {Message}", e.Message);
            throw new InvalidOperationException("Cannot determine active WhatsApp instance", e);
        }
    }

    private async Task<string> BuildZApiUrlAsync(string endpoint)
    {
        var instance = await GetActiveInstanceAsync();
        ValidateInstanceConfiguration(instance);
        return $"https://api.z-api.io/instances/{instance.InstanceId}/{endpoint}";
    }

    private static void ValidateInstanceConfiguration(WhatsAppInstance instance)
    {
        if (string.IsNullOrEmpty(instance.InstanceId))
        {
            throw new InvalidOperationException("Instance ID is required for Z-API activation");
        }
        
        if (string.IsNullOrEmpty(instance.AccessToken))
        {
            throw new InvalidOperationException("Access Token is required for Z-API activation");
        }
    }

    public async Task<ZApiStatus> GetInstanceStatusAsync()
    {
        try
        {
            var url = await BuildZApiUrlAsync("status");
            var request = new HttpRequestMessage(HttpMethod.Get, url);
            AddHeaders(request);

            var response = await _httpClient.SendAsync(request);

            if (response.IsSuccessStatusCode)
            {
                var content = await response.Content.ReadAsStringAsync();
                var responseData = JsonSerializer.Deserialize<Dictionary<string, object>>(content);
                return ParseStatusResponse(responseData ?? new Dictionary<string, object>());
            }
            
            return ZApiStatus.CreateError("Failed to get status");
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error getting Z-API status: {Message}", e.Message);
            return ZApiStatus.CreateError($"Error: {e.Message}");
        }
    }

    public async Task<QrCodeResult> GetQrCodeBytesAsync()
    {
        try
        {
            var url = await BuildZApiUrlAsync("qr-code");
            var request = new HttpRequestMessage(HttpMethod.Get, url);
            AddHeaders(request);

            var response = await _httpClient.SendAsync(request);

            if (response.IsSuccessStatusCode)
            {
                var bytes = await response.Content.ReadAsByteArrayAsync();
                return QrCodeResult.CreateSuccess(bytes, "bytes");
            }
            
            return QrCodeResult.CreateError("Failed to get QR code bytes");
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error getting QR code bytes: {Message}", e.Message);
            return QrCodeResult.CreateError($"Error: {e.Message}");
        }
    }

    public async Task<QrCodeResult> GetQrCodeImageAsync()
    {
        try
        {
            var url = await BuildZApiUrlAsync("qr-code/image");
            _logger.LogInformation("Getting QR code from URL: {Url}", url);

            var request = new HttpRequestMessage(HttpMethod.Get, url);
            AddHeaders(request);

            var response = await _httpClient.SendAsync(request);
            var content = await response.Content.ReadAsStringAsync();
            _logger.LogInformation("Z-API QR Code response: status={StatusCode}, body={Body}", 
                response.StatusCode, content);

            if (response.IsSuccessStatusCode && !string.IsNullOrEmpty(content))
            {
                var responseData = JsonSerializer.Deserialize<Dictionary<string, object>>(content);
                if (responseData?.TryGetValue("value", out var valueObj) == true)
                {
                    var fullValue = valueObj?.ToString();
                    _logger.LogInformation("Full value received: {Value}", 
                        fullValue?.Length > 100 ? $"{fullValue[..100]}..." : fullValue);

                    // Extract base64 part from data URL format (data:image/png;base64,XXXX)
                    string? base64Image = null;
                    if (!string.IsNullOrEmpty(fullValue) && fullValue.StartsWith("data:image/png;base64,"))
                    {
                        base64Image = fullValue["data:image/png;base64,".Length..];
                    }

                    _logger.LogInformation("Base64 image extracted: {Status}", 
                        base64Image != null ? $"present ({base64Image.Length} chars)" : "null");
                    return QrCodeResult.CreateSuccess(base64Image, "base64");
                }
            }

            _logger.LogWarning("Failed to get QR code: status={StatusCode}, body={Body}", 
                response.StatusCode, content);
            return QrCodeResult.CreateError("Failed to get QR code image");
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error getting QR code image: {Message}", e.Message);
            return QrCodeResult.CreateError($"Error: {e.Message}");
        }
    }

    public async Task<PhoneCodeResult> GetPhoneCodeAsync(string phoneNumber)
    {
        try
        {
            var url = await BuildZApiUrlAsync($"phone-code/{phoneNumber}");
            var request = new HttpRequestMessage(HttpMethod.Get, url);
            AddHeaders(request);

            var response = await _httpClient.SendAsync(request);

            if (response.IsSuccessStatusCode)
            {
                var content = await response.Content.ReadAsStringAsync();
                var responseData = JsonSerializer.Deserialize<Dictionary<string, object>>(content);
                
                if (responseData?.TryGetValue("code", out var codeObj) == true)
                {
                    var code = codeObj?.ToString();
                    return PhoneCodeResult.CreateSuccess(code ?? string.Empty, phoneNumber);
                }
            }
            
            return PhoneCodeResult.CreateError("Failed to get phone code");
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error getting phone code: {Message}", e.Message);
            return PhoneCodeResult.CreateError($"Error: {e.Message}");
        }
    }

    public async Task<bool> RestartInstanceAsync()
    {
        try
        {
            var url = await BuildZApiUrlAsync("restart");
            var request = new HttpRequestMessage(HttpMethod.Get, url);
            AddHeaders(request);

            var response = await _httpClient.SendAsync(request);
            return response.IsSuccessStatusCode;
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error restarting instance: {Message}", e.Message);
            return false;
        }
    }

    public async Task<bool> DisconnectInstanceAsync()
    {
        try
        {
            var url = await BuildZApiUrlAsync("disconnect");
            var request = new HttpRequestMessage(HttpMethod.Get, url);
            AddHeaders(request);

            var response = await _httpClient.SendAsync(request);
            return response.IsSuccessStatusCode;
        }
        catch (Exception e)
        {
            _logger.LogError(e, "Error disconnecting instance: {Message}", e.Message);
            return false;
        }
    }

    private void AddHeaders(HttpRequestMessage request)
    {
        var clientToken = _configuration["ZApi:ClientToken"];
        request.Headers.Add("client-token", clientToken);
        request.Headers.Add("Accept", "application/json");
    }

    private static ZApiStatus ParseStatusResponse(Dictionary<string, object> response)
    {
        var connected = ParseBoolean(response.GetValueOrDefault("connected"));
        var smartphoneConnected = ParseBoolean(response.GetValueOrDefault("smartphoneConnected"));
        var session = ParseString(response.GetValueOrDefault("session"));

        return new ZApiStatus
        {
            Connected = connected,
            Session = session,
            SmartphoneConnected = smartphoneConnected,
            NeedsQrCode = !connected,
            RawResponse = response
        };
    }

    private static bool ParseBoolean(object? value)
    {
        return value switch
        {
            bool boolValue => boolValue,
            string stringValue => string.Equals(stringValue, "true", StringComparison.OrdinalIgnoreCase),
            _ => false
        };
    }

    private static string? ParseString(object? value)
    {
        return value?.ToString();
    }
}